packages/commit-parser/parse.ts

```diff
@@ -0,0 +1,175 @@
+import {CommitRoot} from "@romefrontend/ast";
+import {descriptions} from "@romefrontend/diagnostics";
+import {Number0, ob1Add} from "@romefrontend/ob1";
+import {
+	ParserOptions,
+	TokenValues,
+	createParser,
+} from "@romefrontend/parser-core";
+
+import {Symbols, Tokens} from "./types";
+
+export const createCommitParser = createParser((ParserCore) =>
+	class CommitParser extends ParserCore<Tokens> {
+		constructor(opts: ParserOptions) {
+			super(opts, "parse/commit", {});
+			this.ignoreWhitespaceTokens = false;
+		}
+
+		tokenize(index: Number0): undefined | TokenValues<Tokens> {
+			if (
+				this.getInputCharOnly(index) === Symbols.Space ||
+				this.getInputCharOnly(index) === Symbols.Tab
+			) {
+				while (
+					this.getInputCharOnly(index) === Symbols.Space ||
+					this.getInputCharOnly(index) === Symbols.Tab
+				) {
+					index = ob1Add(index, 1);
+				}
+				return this.finishToken("Whitespace", index);
+			}
+
+			if (this.getInputCharOnly(index) === "(") {
+				return this.finishToken("LeftParen");
+			}
+
+			if (this.getInputCharOnly(index) === ")") {
+				return this.finishToken("RightParen");
+			}
+
+			if (this.getInputCharOnly(index) === "!") {
+				return this.finishToken("Exclamation");
+			}
+
+			if (this.getInputCharOnly(index) === ":") {
+				return this.finishToken("Colon");
+			}
+
+			return this.finishValueToken("Word", this.getInputCharOnly(index));
```

Comment:
I think this reads a lot clearer with a `switch`.

```suggestion
			consr char = this.getInputCharOnly(index);
			switch (char) {
				case Symbols.Space:
				case Symbols.Tab: {
					while (
						char === Symbols.Space ||
						char === Symbols.Tab
					) {
						index = ob1Add(index, 1);
					}
					return this.finishToken("Whitespace", index);
				}

				case "(":
					return this.finishToken("LeftParen");

				case ")":
					return this.finishToken("RightParen");

				case "!":
					return this.finishToken("Exclamation");

				case ":":
					return this.finishToken("Colon");

				default:
					return this.finishValueToken("Word", char);	
			}
```